google/cloud/pubsublite/internal/partition_publisher.h

```diff
@@ -0,0 +1,322 @@
+          });
+    }
+    root.set_value();
+  }
+
+  void Read() {
+    promise<void> root;
+    future<void> root_future = root.get_future();
+    {
+      std::lock_guard<std::mutex> g{mu_};
+      if (shutdown_) return;
+      root_future.then(ChainFuture(resumable_stream_->Read()))
+          .then([this](future<absl::optional<MessagePublishResponse>>
+                           optional_response_future) {
+            auto optional_response = optional_response_future.get();
+            if (!optional_response) Read();
+            Batch batch;
+            {
+              std::lock_guard<std::mutex> g{mu_};
+              if (shutdown_) return;
+              // unlikely order of events
+              // if we have `Read()` (pending) -> `Write()` (completes
+              // successfully) -> `Write()` (fails) -> previous `Read`
+              // successfully completes -> `initializer_` grabs locks and
+              // rebatches and unlocks -> below `Read` continuation grabs lock
+              // and tries to satisfy future in nonexistent first element of
+              // `in_flight_batches_`
+              if (!in_flight_batches_.empty()) {
+                batch = std::move(in_flight_batches_.front());
+                in_flight_batches_.pop_front();
+              }
+            }
+            size_t offset = optional_response->start_cursor().offset();
+            for (auto& message_with_future : batch) {
+              Cursor c;
+              c.set_offset(offset);
+              ++offset;
+              message_with_future.message_promise.set_value(std::move(c));
+            }
+            Read();
+          });
+    }
+    root.set_value();
+  }
+
+  void SatisfyOutstandingMessages(Status const& status) {
+    std::vector<MessageWithFuture> messages_with_futures;
+    {
+      std::lock_guard<std::mutex> g{mu_};
+      for (auto& batch : in_flight_batches_) {
+        for (auto& message_with_future : batch) {
+          messages_with_futures.push_back(std::move(message_with_future));
+        }
+      }
+      in_flight_batches_.clear();
+      for (auto& batch : batches_) {
+        for (auto& message_with_future : batch) {
+          messages_with_futures.push_back(std::move(message_with_future));
+        }
+      }
+      batches_.clear();
+      for (auto& message_with_future : unbatched_messages_with_futures_) {
+        messages_with_futures.push_back(std::move(message_with_future));
+      }
+      unbatched_messages_with_futures_.clear();
+    }
+    for (auto& message_with_future : messages_with_futures) {
+      message_with_future.message_promise.set_value(StatusOr<Cursor>(status));
+    }
+  }
+
+  void DestroyCancelToken() { auto cancel_token = std::move(cancel_token_); }
+
+  // When initializer is called, no outstanding Read() or Write() futures will
+  // succeed.
+  StreamInitializer<MessagePublishRequest, MessagePublishResponse>
+      initializer_ = [this](UnderlyingStream stream) {
+        Rebatch(true);
```

Comment:
This isn't everything that needs to happen. You also need to send the initial request and wait for the initial response.